[Security Solution][Alert Details] - push and overlay expandable flyout

PhilippeOberti opened the pull request:
## Summary

This PR adds a new functionality to the `kbn-expandable-flyout` package and its usage in the Security Solution application.

The package's flyout now support to be rendered in `overlay` or `push` mode, following [EUI's recommendation](https://eui.elastic.co/#/layout/flyout#push-versus-overlay).
A gear icon button is rendered in the top right corner, next to the close button. When clicked, a menu appears where users can select `push` or `overlay` values. `overlay` is the default value. If `push` is selected, a `Reset to default` empty button can be used to reset to `overlay`.

Overlay option selected (by default)
![Screenshot 2024-09-04 at 12 10 34 PM](https://github.com/user-attachments/assets/87f57238-9b44-4d29-9516-9eb329c49bb2)

Push option selected
![Screenshot 2024-09-04 at 12 10 42 PM](https://github.com/user-attachments/assets/80e7879a-b238-46ba-9c13-2c8e236e138f)

The flyout should be toggled between `overlay` and `push` mode in all the pages it's been currently used in:
- alerts page
- rule creation page
- explore pages (host, users...)
- case detail page

https://github.com/user-attachments/assets/b4cec138-802c-430d-8f37-01258e6afef3

But the flyout cannot be set to `push` mode when opened  from Timeline. Timeline is a modal (an EUI Portal to be precise), and getting the portal as well as the overlay mask to correctly resize according to the flyout's width (which is dynamic, changes with the screen size and also changes if the flyout is in collapsed or expanded mode) is very difficult.
A future PR might add this functionality to TImeline. At this time, the flyout offers the option to disable the `push/overlay` toggle as well as an icon to show a tooltip to explain why.

https://github.com/user-attachments/assets/e00961c8-cc75-4eb9-b34d-544bc4391d5c

#### Notes

The package also offers a way to hide the gear icon entirely. In the future, we might need a bit more flexibility if we want to be able to show the gear icon with options others than the `push/overlay` entry.

Finally the state of the flyout type (`overlay` vs `push`) is saved in local storage so that users don't have to set the value over and over again. This state is persisted within the `kbn-expandable-flyout` package so that developers don't have to worry about setting it up. The package uses its internal `urlKey` to guarantee that the key used to save in localStorage is unique. This means that `memory` flyouts cannot persist the `push` or `overlay` states, this is expected.

https://github.com/elastic/kibana/assets/17276605/500315b5-07d4-4498-aab9-ee2e2be0253b

### Notes

The package's README has been updated.
New Storybook stories have been added to reflect the new push/overlay functionality.

https://github.com/elastic/kibana/issues/182593

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios **_TODO_**
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))

PhilippeOberti committed:
[Security Solution][Alert Details] - add support for push/overlay to kbn-expandable-flyout package

PhilippeOberti commented:
buildkite test this

elasticmachine commented:
Pinging @elastic/security-threat-hunting-investigations (Team:Threat Hunting:Investigations)

christineweng commented on packages/kbn-expandable-flyout/src/components/render_menu.tsx:
```diff
@@ -0,0 +1,153 @@
+    setPopover(!isPopoverOpen);
+  };
+
+  const panels = [
```
what do you think about moving the panels to closer to the alert flyout code? This looks too specific to be in the package for me.. i could imagine UI changes or more configs down the road, and we'd have to modify the package... like if we want to add a tooltip that says (push is disabled for timeline), we have to do it here?

PhilippeOberti replied:
I went back and forth on this a few times... I thought about the tooltip you're mentioning and that almost tipped the balance towards having this code outside of the package, but ultimately, what I like about having it inside was:
- we control the UI (so we don't end up with n versions of a way to do push/overlay and all the other options we'll provide)
- this really seems to me like a package code, as it directly touches how the packages works internally. By that I mean that if for some reason we decide to switch `EuiFlyout` to something else, we don't want developers to have to know the name of the property to use, or even how it works... Granted that they kinda have to know today because they can pass the type (`'push'` or `'overlay'`), but it's typed and kept to a minimum...

My philosophy was let's keep it tightly coupled with the package for now, and when people complain about a limitation with a design that they have, we should be able to easily extract it into a reusable component, and make the popover in the package render whatever is passed in...

Happy to discuss this more though, great comment!

christineweng replied:
sure, we can discuss more and brainstorm together. My main concern is the setting menu currently is not very flexible:
- ideally we should be able pass the `panels` and decide what settings to show, i.e. host or alert specific settings
- if other devs don't care about push vs overlay, the setting is still showing
- all the future config will sit in the package, it will be a pain to manage if we have to review settings outside of investigations

However, it is nice to have some package level control out of the box, so i understand why you take this approach. 

If no additional settings are planned for the next cycle or two, it's probably okay. just don't want to see the work here reverted in a couple of months :)

PhilippeOberti replied:
I actually just tried the opposite way, where I move the entirety of the code within the package, and the live only the decision to enable/disable the functionality to the security solution (or other) plugin(s).

I like this approach, but I still see valid point of doing it like you suggested. Let's discuss this this week!

michaelolo24 replied:
Yea, I agree with Christine here. Consumers of the flyout package may want it to behave more like `pushBreakpoint={800} // innerWidth of window at which point the flyout enters "push" mode` rather than have a toggleable push vs overlay option. As an aside, I figure design at some point will want some kind of icon for defining push vs overlay if that's the plan as well. Something like one of these. When push is enabled, the icon is highlighted/depressed, when it's disabled it looks normal...

![image](https://github.com/user-attachments/assets/5020a942-1ff4-49ab-a9e6-86db6814ae77)

PhilippeOberti replied:
I don't think the `pushBreakpoint={800}` approach you're suggestion would make sense for us right now, because at this time there is a single flyout for the entire Security Solution application. While it could make sense to have for the threshold to be at 800 on the alerts page, maybe on the rule details or the host or user pages it doesn't make any sense at all and we would want a completely different value...
I like the approach I'm taking in the second commit because now the functionality is fully internal to the package. Developers don't have to know about how it works, how to save it...

Also having the threshold customizable makes little sense to me if you combine the push/overlay functionality for a resizable flyout ([coming shortly](https://github.com/elastic/kibana/pull/182614) after this PR). If the flyout is resizable, I think it makes more sense to just let the user toggle on/off the push vs overlay. It would be much more complex (in code and in the UI) to allow to manage resizable as well as a `pushBreakPoint` value. Would we want those to be different?

I understand that the current approach is different from Discover which manages push/overlay by the window width, but they also have a single page their flyout is accessible from. We have the entire Security Solution application...
Happy to discuss this more though because I'm sure I'm not seeing the whole picture!

michaelolo24 replied:
Just documenting that we spoke offline about this. To re-iterate your arguments here make sense especially with the introduction of the resizable flyout. Providing users the flexibility to choose the option makes the most sense, and allowing developers to just enable/disable the functionality also makes sense. Looking forward, if we were to enable a "kiosk" mode then the "flyout type" will probably be  `overlay`, `push`, `modal / kiosk / something`. Can we rename "flyout type" to "layout type" or "layout setting" or just "layout"... When I hear flyout type, I think alert vs event vs entity, etc...

PhilippeOberti replied:
I decided to keep flyout `type` for now for 2 reasons:
- while for us in SecuritySolution the word `type` could mean signal, alert, event... it wouldn't mean that for anyone else. This code leaves in the `packages` folder, so far enough I think from Security Solution
- more importantly, `type` is what EUI uses everywhere in the code and in their docs. It makes a lot of sense to use their terminology, because that's directly what we're using

PhilippeOberti committed:
moving all the code within the package

PhilippeOberti committed:
- add option to hide the entire settings menu
- add unit tests
- add storybook stories
- update README

christineweng reviewed (approved):
Code changes LGTM! Had some comments on the visuals but neither is a blocker

- It's not noticeable when in normal screen width, but with a wider monitor,  some of the buttons are pushed to the nav side panel. Not sure this can be fixed though(?) 
![image](https://github.com/user-attachments/assets/358bf10a-d48b-47d9-814b-0432997b75a7)

- When `Push` is selected, there is a reset to default button, but it just set it to overlay. I'm not sure it's worth 'resetting' as there are only 2 options?
![image](https://github.com/user-attachments/assets/3a4a24a6-5464-4575-82bf-93b88f758893)

PhilippeOberti committed:
- add disable push/overlay type back
- add more storybook stories
- remove reset push/overlay option
- add info icon to explain why option is disabled

PhilippeOberti commented on packages/kbn-expandable-flyout/src/components/render_menu.test.tsx:
```diff
@@ -0,0 +1,63 @@
+    const flyoutTypeProps = {
+      type: 'overlay' as EuiFlyoutProps['type'],
+      onChange: jest.fn(),
+      disabled: false,
```
I have actually added it back after talking to Paul

PhilippeOberti commented:
> Code changes LGTM! Had some comments on the visuals but neither is a blocker
> 
> * It's not noticeable when in normal screen width, but with a wider monitor,  some of the buttons are pushed to the nav side panel. Not sure this can be fixed though(?)

I feel like this is an issue with the alerts page. While it is true that this new functionality reveals the bug more than before, it does not introduce it. This would happen (if I'm not mistaken) on a very very very small screen width as well.
Also I don't really want to have this PR to modify too many other places. It would mean a much higher scope, more people reviewing for something that's not completely related...

> * When `Push` is selected, there is a reset to default button, but it just set it to overlay. I'm not sure it's worth 'resetting' as there are only 2 options?

Very good point, you're not the first person mentioning that. I have removed the reset option entirely!

nastasha-solomon commented on packages/kbn-expandable-flyout/src/components/render_menu.tsx:
```diff
@@ -0,0 +1,161 @@
+                id: 'overlay',
+                label: FLYOUT_TYPE_OVERLAY_MODE,
+                'data-test-subj': RENDER_MENU_FLYOUT_TYPE_BUTTON_GROUP_OVERLAY_TEST_ID,
+                toolTipContent: 'Shows the flyout on top of the page content',
```
@PhilippeOberti how's this:

```suggestion
                toolTipContent: 'Displays the flyout over the page',
```

PhilippeOberti replied:
that works for me! Thank you so much for the quick feedback!

nastasha-solomon commented on packages/kbn-expandable-flyout/src/components/render_menu.tsx:
```diff
@@ -0,0 +1,161 @@
+                id: 'push',
+                label: FLYOUT_TYPE_PUSH_MODE,
+                'data-test-subj': RENDER_MENU_FLYOUT_TYPE_BUTTON_GROUP_PUSH_TEST_ID,
+                toolTipContent: 'Squishes the page content to show the flyout next to it',
```
```suggestion
                toolTipContent: 'Displays the flyout next to the page',
```
